### PR TITLE
refactor(codex-github): whole-body exact-template verdict classifier (#1491)

### PR DESCRIPTION
**Pull request by lhpaul**
## Summary

Implements issue #1491's implementation plan (`docs/specs/developments/20260817203204_1491-conservative-verdict-classifier/2_1491-conservative-verdict-classifier_implementation-plan.md`, merged `a0b8f8c7`).

`codex_response_is_approved` in `codex-github-reviewer.sh` is rewritten from an open-ended negated-approval-vocabulary block-list to a whole-body, whitespace-normalized **exact match** against `CODEX_APPROVED_TEMPLATES` — a small set of clean-response templates captured verbatim from real Codex responses, each including the complete vendor `<details>` "About Codex in GitHub" footer. There is no truncation step: the footer is part of the literal being matched, not discarded before comparison, closing Codex GitHub finding `3803545669` (the finding that motivated this revision).

`codex_response_is_blocking` is unchanged (confirmed byte-identical to the merge base by function-range diff). Decision 6 gates the acknowledgement branch at all four verdict sites (main loop, async-arrival, async-final, async-reaction-final) on the evidence being non-terminal, so a footer-bearing near-miss on terminal evidence always safe-fails instead of being misrouted to a wait/timeout.

## Wire-format re-verification (Protocol 02 implementation-start check, mandated by the plan)

Re-fetched live Codex responses from PRs #1489, #1490, and #1492 before any edit:

- **Clean-response verdict sentence**: PR #1489's root comment still matches `CODEX_APPROVED_TEMPLATES`' entry exactly (verified by running the actual regex against the live capture).
- **Vendor `<details>` footer**: byte-identical (modulo one API-transport-only trailing newline, absorbed by whitespace normalization) across the PR #1489 root comment and every PR #1490/#1492 review body.

**Result: matched, no drift.** No stop-and-report was needed.

## Test disposition (derived fresh against the file, not carried forward from any prior revision)

- **Group APPROVED** (8 members): full body replacement with the exact template, preserving each scenario's distinguishing mock sequencing (2 template-anchored + 6 routing-testing scenarios: `codex_reaction_with_current_review`, `codex_reaction_then_late_review`, `codex_async_reaction_then_late_review`, `codex_main_loop_env_then_newer_review_supersedes`, `codex_latest_current_review`, `codex_environment_with_current_review`).
- **Group RETARGETED** (19 members): `VERDICT: APPROVED` → `VERDICT: NEEDS_REVISION (unrecognized response format — safe-fail)`, paired `exit_clean` `0` → `1`. 8 scenarios whose names asserted `stays_approved`/`_approved_` were renamed to `_safe_fails_` so the name never disagrees with the expectation. `codex_usage_limit_topic_mention_not_quota`'s competing root-comment fixture was also updated to the new exact template (Codex GitHub finding `3805611400`) so the tie-break still isolates whether the review is correctly classified rather than both sides collapsing into the same "unrecognized" tier.
- **Two additional fixtures found by executing the full suite** (outside the 27 `VERDICT: APPROVED`-asserting scenarios, so not caught by that enumeration): `codex_main_loop_env_then_review` (the `SEEN_ENVIRONMENT_ERROR`-supersede check lives inside the `is_approved` branch) and `codex_tied_usage_limit_with_approval_phrase` (an unrecognized body now outranks usage-limit in `codex_response_priority`, so the clean side of the tie must stay a real template).
- **New scenarios**: Parser-risk addendum E1, E2, E3, E5–E13, E15–E24 (E4 and E14 already covered by existing fixtures), including the real, live-fetched PR #1489/#1490 bodies, plus four `codex_footer_near_miss_*_safe_fails` scenarios — one per verdict site — proving Decision 6's gate is independently correct at each of its four hand-edited copies (verified by reverting only one site's gate in a scratch copy of the script and confirming only that scenario regressed to `TIMED_OUT`, the other three still passed).

## Verification

- `bash scripts/development-workflow/tests/test-pr-review-loop.sh`: **684 passed, 0 failed**.
- Comment-filtered deletion-list search for every symbol this plan or its prior revisions ever introduced: returns nothing.
- Five (not four) real call sites of `codex_response_is_approved` confirmed (the four verdict sites plus one inside `codex_response_priority`).
- `codex_strip_not_only_idiom`: comment-filtered count = 2 (definition + one call, inside `codex_response_is_blocking`).
- Merge-base (`a0b8f8c7d9b63ea0f2ace945f78990579ceca828`) function-range diff of `codex_response_is_blocking`, plus the `CODEX_BLOCKING_PATTERN`/`CODEX_MERGE_REFUSAL_PATTERN`/`CODEX_NEGATION_WORDS` definitions: **empty (byte-identical)**.
- Live-refetched PR #1489 body approves end-to-end through the real script; live-refetched PR #1490 review still safe-fails.
- All Parser-risk addendum edge cases (E14–E24) verified against the real script.
- Smoke test runbook (`docs/testing/workflow/1491-conservative-verdict-classifier.smoke-test.md`) walked in full — all 15 steps Pass, results recorded in the runbook.
- Lint: `markdownlint-cli2` (0 errors), `markdown-heuristic-lint.py` (exit 0), `check-changelog-duplicate-headers.sh` (passed), `workflow-shell-snippet-lint.py --base-ref origin/develop` (exit 0), `workflow-shell-guard-lint.py --base-ref origin/develop` (exit 0), `shellcheck --severity=warning` on both changed `.sh` files (only pre-existing, untouched warnings at unrelated lines).

## Self-review

Reviewed `git diff develop...HEAD`: no stale markers, no debug/TODO leftovers, no orphaned "stays_approved" variable-name remnants after the 8 scenario renames, no duplicate symbol definitions. `pr-review-loop.sh`, `AGENTS.md`, and `REVIEW.md` are untouched, matching the plan's explicit "no change" calls. The plan's own Decision-gate matrix (verdict strings, exit codes) is unchanged by this implementation — no new gate outcomes were introduced.

## Scope

Everything the plan specifies, nothing more. No changes to `pr-review-loop.sh` beyond what the classifier redesign strictly required (none were required — the redesign is entirely inside `codex-github-reviewer.sh`).

Closes #1491.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by lhpaul** [comment]
## Protocol 91 Step 7a — Internal Review (Claude, implementation stage)

**Verdict: APPROVED** (after one fix applied and pushed: `a2a20e7b`)

Reviewed against the merged implementation plan (`docs/specs/developments/20260817203204_1491-conservative-verdict-classifier/2_..._implementation-plan.md`) and `REVIEW.md`'s Code Review Checklist. Design Review Gate skipped (Path A, no frontend files). Repository mode: `single_repo` — this repository owns the artifact.

### Findings

1. **[blocking, fixed]** All 19 Group RETARGETED scenarios' paired exit-code assertions retained the `_exit_clean` name suffix after their expected value was correctly retargeted from `"0"` to `"1"` (e.g. `codex_usage_limit_topic_mention_not_quota_exit_clean` asserting `"1"`). The plan's own Test disposition section requires renaming any assertion "whose name now asserts the opposite of its expectation... so the name and expectation never disagree on develop" — this was applied to base scenario names (`_stays_approved_` → `_safe_fails_`) but missed on the paired exit-assertion identifier. Fixed in `a2a20e7b`: renamed all 19 to `_exit_needs_revision`, matching the file's own established convention (e.g. `codex_shouldnt_be_merged_blocking_root_comment_exit_needs_revision`). Full suite re-verified clean after the fix (684 passed, 0 failed).

No other findings — the implementation matches the plan precisely everywhere I checked (see scrutiny points below), and lint gates are clean.

### Scrutiny points 1–9 (evidence)

1. **Decision 6 gate at all four verdict sites** — read each site directly. Main-loop and async-arrival use the positive form `[ SOURCE != "review" ] && grep ...` (wait branch, correctly gated to non-terminal only). Async-final and async-reaction-final use the negated form `[ SOURCE = "review" ] || ! grep ...` (safe-fail branch, correctly unconditional for terminal evidence). All four confirmed correct by direct code reading.
2. **`codex_response_is_blocking` byte-identical to merge base** — extracted the full function body (open brace to close brace) from both `a0b8f8c7` and `HEAD`, `diff` returned empty (43/43 lines identical). `codex_strip_not_only_idiom`'s call inside it confirmed present.
3. **Five call sites of `codex_response_is_approved`** — confirmed by grep: definition (1) + four verdict sites + one inside `codex_response_priority` = 5, matching the plan's correction (finding `3805786163`).
4. **Template/footer byte-exact, SHA bound `[0-9a-f]{7,40}`** — diffed the shipped `CODEX_APPROVED_TEMPLATES` array literal against the plan's Code Samples literal: identical. `codex_strip_codex_footer`/`CODEX_FOOTER_OPENING_LITERAL` confirmed absent (comment-filtered grep, no matches).
5. **Deleted symbols actually gone** — comment-filtered grep for the full deletion list (`CODEX_APPROVAL_PATTERN`, `CODEX_NEGATED_APPROVAL*`, `CODEX_CLEAN_SIGNAL*`, `CODEX_APPROVAL_(NEGATION|HEDGE|ACTIONABLE|DISQUALIFIER)*`, `CODEX_RESIDUE_FILLER*`, `CODEX_VENDOR_FLAVOR*`, `codex_excise_clean_signals`, `codex_residue_is_closed_grammar`, `codex_response_first_paragraph`, `codex_strip_vendor_metadata_lines`) returned zero matches.
6. **Four near-miss scenarios independently diagnostic** — built an isolated scratch harness (git worktree + minimal extracted test runner) and reverted only the async-final site's gate to its true pre-Decision-6 form (verified against the actual merge-base code first, since my first attempt at the revert had the wrong polarity). Result: only `codex_footer_near_miss_async_final_safe_fails` regressed to `VERDICT: TIMED_OUT`; the other three (main-loop, async-arrival, async-reaction-final) still passed. Matches the developer's claimed verification exactly.
7. **Stale comments** — checked all six locations named in the plan (file header, inline "Verdict parsing" duplicate, `codex_response_has_fence_marker`, `codex_strip_quoted_spans`, `codex_strip_not_only_idiom` docstrings, plus `codex_response_is_blocking`'s own rationale comment). All updated correctly; none describes a removed mechanism as current.
8. **Two extra-fixed fixtures** — `codex_main_loop_env_then_review` and `codex_tied_usage_limit_with_approval_phrase` both use exact-template bodies with valid distinct SHAs; their in-file comments correctly explain why the fix was needed even though neither scenario's own assertion mentions `APPROVED`. Regression value preserved.
9. **CHANGELOG** — entry present under `[Unreleased]` → `### Changed`, byte-identical to the plan's literal (modulo list indentation), no trailing whitespace, `check-changelog-duplicate-headers.sh` passes.

### Scope check

Diff matches the plan's file list exactly (6 files). `codex-github.md` gained a new "Verdict Classification" section (no such section existed before — not a duplicate/opportunistic edit). Protocol 93 gained the exact block the plan specifies. `AGENTS.md`, `REVIEW.md`, `.agents/`, `.codex/` untouched, matching the plan's explicit "no change" calls.

### Verification run

- `bash scripts/development-workflow/tests/test-pr-review-loop.sh`: **684 passed, 0 failed** (confirmed independently, both before and after my fix).
- `bash -n` on both changed shell files: clean.
- `shellcheck --severity=warning` on both: only pre-existing SC2034 warnings at test-file lines 1699-1700, confirmed identical at merge base (untouched by this PR).
- `./node_modules/.bin/markdownlint-cli2` (changed docs + CHANGELOG): 0 errors.
- `python3 scripts/lint/markdown-heuristic-lint.py` on changed docs: clean.
- `python3 scripts/lint/workflow-shell-snippet-lint.py --base-ref origin/develop`: exit 0.
- `python3 scripts/lint/workflow-shell-guard-lint.py --base-ref origin/develop`: exit 0.
- `bash scripts/lint/check-changelog-duplicate-headers.sh CHANGELOG.md`: passed.

### Commit pushed

`a2a20e7b` — fix(codex-github): rename retargeted exit assertions

**Comment by lhpaul** [comment]
@codex review (review triggered by workflow runner, commit: a2a20e7b4836)

**Comment by lhpaul** [comment]
### Automated Reviewer Loop Summary

**Result:** clean — no blocking findings
**Platforms:** pr-agent (clean), codex-github (clean)
**Findings:** 0 blocking, 0 suggestions

**Ready reviewer phase:** No net-new blocker was found after the draft GitHub gate.
**Ready-phase platforms:** codex-github

**Advisory findings (non-blocking):**
- Possible Issue ([view comment](https://github.com/lhpaul/ai-dev-framework-template/pull/1494#issuecomment-5333033016))
- Possible Issue ([view comment](https://github.com/lhpaul/ai-dev-framework-template/pull/1494#issuecomment-5333033016))

*Posted automatically by `pr-review-loop.sh`.*

<details>
<summary>Reviewer-loop history (3 iterations)</summary>


```json
{
  "schema": "reviewer_loop_history.v1",
  "pr_number": 1494,
  "updated_at": "2026-08-18T22:39:10Z",
  "history_status": "available",
  "history_unavailable_reason": "",
  "entries": [
    {
      "iteration": 1,
      "recorded_at": "2026-08-18T20:15:58Z",
      "head_sha": "a2a20e7b4836cfb5d20b75e39e01447757434e34",
      "result": "needs_fixes",
      "reason": "unresolved_review_threads",
      "platforms": [
        "pr-agent (clean)",
        "codex-github (needs_fixes)"
      ],
      "blocking_count": 1,
      "suggestion_count": 0,
      "unresolved_thread_count": 0,
      "late_threads_found": 0,
      "phase_after_clean": {
        "enabled": true,
        "platforms": [
          "codex-github"
        ],
        "started": true,
        "net_new_blocker": true,
        "blocking_platform": "codex-github"
      }
    },
    {
      "iteration": 2,
      "recorded_at": "2026-08-18T22:06:13Z",
      "head_sha": "94de701e12b6b1e27179ac9b203990bc83bdd35b",
      "result": "needs_fixes",
      "reason": "unresolved_review_threads",
      "platforms": [
        "pr-agent (clean)",
        "codex-github (needs_fixes)"
      ],
      "blocking_count": 1,
      "suggestion_count": 0,
      "unresolved_thread_count": 0,
      "late_threads_found": 0,
      "phase_after_clean": {
        "enabled": true,
        "platforms": [
          "codex-github"
        ],
        "started": true,
        "net_new_blocker": true,
        "blocking_platform": "codex-github"
      }
    },
    {
      "iteration": 3,
      "recorded_at": "2026-08-18T22:39:10Z",
      "head_sha": "b4655e6a2e92fec9594e1b3b19f94f1da9ca6601",
      "result": "clean",
      "reason": "",
      "platforms": [
        "pr-agent (clean)",
        "codex-github (clean)"
      ],
      "blocking_count": 0,
      "suggestion_count": 0,
      "unresolved_thread_count": 0,
      "late_threads_found": 0,
      "phase_after_clean": {
        "enabled": true,
        "platforms": [
          "codex-github"
        ],
        "started": true,
        "net_new_blocker": false,
        "blocking_platform": ""
      }
    }
  ]
}
```
</details>

**Comment by lhpaul** [comment]
## Follow-up: PR #1494's own Codex review falsified the single-flavor-literal assumption

**Live evidence, found during Step 7a review**: this PR's own first real-world Codex review (comment
[`5333550055`](https://api.github.com/repos/lhpaul/ai-dev-framework-template/issues/comments/5333550055))
returned a clean response reading `:rocket:` instead of the sole evidenced `Swish!` literal, so the shipped
classifier safe-failed on its first real-traffic exercise. Not an implementation bug — the code did exactly
what the plan specified; this is the plan's own top-ranked disclosed risk materializing.

**Human decision**: enumerate the flavor slot as a bounded ERE alternation of every evidenced token (never a
wildcard/character class). Safe direction: an incomplete list produces a false `NEEDS_REVISION`, not a false
`APPROVED`.

**Sweep result**: swept every Codex clean-verdict root comment reachable in this repo's history
(`commenter:chatgpt-codex-connector[bot]` across ~47 PRs). Found **14 distinct flavor tokens**, not the 1-2
anticipated — a discovery rate suggesting the true pool is larger still:

`Swish!`, `:rocket:`, `Nice work!`, `Chef's kiss.`, `You're on a roll.`, `:tada:`, `Another round soon,
please!`, `:+1:`, `Bravo.`, `Keep it up!`, `Delightful!`, `Keep them coming!`, `Can't wait for the next one!`,
`More of your lovely PRs please.`

Full provenance table (PR + comment id per token) is in the plan's new **Decision 2 Addendum**.

**Implemented**: `CODEX_APPROVED_TEMPLATES`' flavor slot is now `(Swish!|:rocket:|Nice work!|...)`, each token
ERE-escaped as a literal. 14 new regression scenarios added (one per token, plus one proving an unevidenced
token still safe-fails). Plan, smoke-test runbook, and `codex-github.md` all updated.

**Verified**:
- Real `:rocket:` body (PR #1494, comment `5333550055`) → `APPROVED`
- Real `Swish!` body (PR #1489, comment `5294701769`) → `APPROVED`
- Unevidenced flavor token → `NEEDS_REVISION`
- All prior E1–E24 cases unchanged
- Full suite: **712 passed, 0 failed** (684 + 28 new assertions)
- All lint gates clean

**Commit**: `be3e945c`

**Comment by lhpaul** [comment]
@codex review (review triggered by workflow runner, commit: 94de701e12b6)

**Comment by lhpaul** [comment]
## Planted-violation transcripts (addresses review finding `3808305143`)

Per `REVIEW.md`'s Verification Discipline / Code Review Checklist → Pass 2 → "PRs that add or modify an automated check": each materially modified check below is proven by mutating the check at a cited file:line, running the exact command, capturing the failing output, restoring, and capturing the passing output. All transcripts run from a `git worktree` at commit `94de701e` (this PR's head at the time of this proof) via the isolated harness convention already used in `test-pr-review-loop.sh`.

**A real, pre-existing test-isolation gap was found and fixed while producing these transcripts**: the main-loop and async-arrival Decision-6 near-miss scenarios' mocks returned the near-miss body on *every* subsequent poll call, so a single-site mutation to either gate was silently rescued by a downstream site's still-correct gate — the transcript below for main-loop shows this happening on the first attempt. Both mocks are now call-counted (near-miss body on exactly one call, empty afterward) to match the isolation convention `async-final`/`async-reaction-final` already used, and are now genuinely independent per-site tests. This fix is included in the commit referenced below.

---

### 1. Decision 6 gate — main-loop, `codex-github-reviewer.sh:1605`

Mutation: remove the `[ "$BOT_RESPONSE_SOURCE" != "review" ] &&` condition, reverting to the unconditional pre-Decision-6 form.

```diff
- elif [ "$BOT_RESPONSE_SOURCE" != "review" ] && grep -qi "If Codex has suggestions, it will comment; otherwise it will react with" <<< "$BOT_RESPONSE_FULL"; then
+ elif grep -qi "If Codex has suggestions, it will comment; otherwise it will react with" <<< "$BOT_RESPONSE_FULL"; then
```

**First attempt, before the isolation fix — no visible failure (the rescue bug):**

```
$ bash <isolated-harness-for-codex_footer_near_miss_main_loop_safe_fails>
PASS: codex_footer_near_miss_main_loop_safe_fails_exit_needs_revision
PASS: codex_footer_near_miss_main_loop_safe_fails_verdict
PASS: codex_footer_near_miss_main_loop_safe_fails_resolves_at_main_loop
Tests: 3 passed, 0 failed
```

This is why the mock isolation fix above was needed. **After the fix, baseline (violation absent):**

```
$ bash <fixed-isolated-harness>
PASS: codex_footer_near_miss_main_loop_safe_fails_exit_needs_revision
PASS: codex_footer_near_miss_main_loop_safe_fails_verdict
PASS: codex_footer_near_miss_main_loop_safe_fails_resolves_at_main_loop
Tests: 3 passed, 0 failed
```

**Failing run (violation present at `codex-github-reviewer.sh:1605`, fixed mock):**

```
FAIL: codex_footer_near_miss_main_loop_safe_fails_exit_needs_revision — expected '1', got '2'
FAIL: codex_footer_near_miss_main_loop_safe_fails_verdict — expected 'VERDICT: NEEDS_REVISION (unrecognized response format — safe-fail)', got 'VERDICT: TIMED_OUT — no response from 'chatgpt-codex-connector[bot]' after 1s (budget 1s) across up to 1 attempt(s)'
PASS: codex_footer_near_miss_main_loop_safe_fails_resolves_at_main_loop
Tests: 1 passed, 2 failed
```

**Restored, passing run:**

```
PASS: codex_footer_near_miss_main_loop_safe_fails_exit_needs_revision
PASS: codex_footer_near_miss_main_loop_safe_fails_verdict
PASS: codex_footer_near_miss_main_loop_safe_fails_resolves_at_main_loop
Tests: 3 passed, 0 failed
```

Scenario that detects the violation: `codex_footer_near_miss_main_loop_safe_fails` (the `_exit_needs_revision` and `_verdict` assertions).

---

### 2. Decision 6 gate — async-arrival, `codex-github-reviewer.sh:1823`

Mutation: same shape as above.

```diff
- elif [ "$ASYNC_BOT_RESPONSE_SOURCE" != "review" ] && grep -qi "If Codex has suggestions, it will comment; otherwise it will react with" <<< "$ASYNC_BOT_RESPONSE_FULL"; then
+ elif grep -qi "If Codex has suggestions, it will comment; otherwise it will react with" <<< "$ASYNC_BOT_RESPONSE_FULL"; then
```

**Baseline (violation absent, fixed mock):**

```
PASS: codex_footer_near_miss_async_arrival_safe_fails_exit_needs_revision
PASS: codex_footer_near_miss_async_arrival_safe_fails_verdict
PASS: codex_footer_near_miss_async_arrival_safe_fails_resolves_at_async_arrival
Tests: 3 passed, 0 failed
```

**Failing run (violation present at `codex-github-reviewer.sh:1823`):**

```
FAIL: codex_footer_near_miss_async_arrival_safe_fails_exit_needs_revision — expected '1', got '2'
FAIL: codex_footer_near_miss_async_arrival_safe_fails_verdict — expected 'VERDICT: NEEDS_REVISION (unrecognized response format — safe-fail)', got 'VERDICT: TIMED_OUT — no response from 'chatgpt-codex-connector[bot]' after 1s (budget 1s) across up to 1 attempt(s)'
PASS: codex_footer_near_miss_async_arrival_safe_fails_resolves_at_async_arrival
Tests: 1 passed, 2 failed
```

**Restored, passing run:**

```
PASS: codex_footer_near_miss_async_arrival_safe_fails_exit_needs_revision
PASS: codex_footer_near_miss_async_arrival_safe_fails_verdict
PASS: codex_footer_near_miss_async_arrival_safe_fails_resolves_at_async_arrival
Tests: 3 passed, 0 failed
```

Scenario that detects the violation: `codex_footer_near_miss_async_arrival_safe_fails`.

---

### 3. Decision 6 gate — async-final, `codex-github-reviewer.sh:1929`

Mutation: remove the `[ "$ASYNC_FINAL_BOT_RESPONSE_SOURCE" = "review" ] ||` disjunct from the negated-form gate.

```diff
- elif [ "$ASYNC_FINAL_BOT_RESPONSE_SOURCE" = "review" ] || ! grep -qi "If Codex has suggestions, it will comment; otherwise it will react with" <<< "$ASYNC_FINAL_BOT_RESPONSE_FULL"; then
+ elif ! grep -qi "If Codex has suggestions, it will comment; otherwise it will react with" <<< "$ASYNC_FINAL_BOT_RESPONSE_FULL"; then
```

**Baseline (violation absent):**

```
PASS: codex_footer_near_miss_async_final_safe_fails_exit_needs_revision
PASS: codex_footer_near_miss_async_final_safe_fails_verdict
PASS: codex_footer_near_miss_async_final_safe_fails_resolves_at_async_final
Tests: 3 passed, 0 failed
```

**Failing run (violation present at `codex-github-reviewer.sh:1929`):**

```
FAIL: codex_footer_near_miss_async_final_safe_fails_exit_needs_revision — expected '1', got '2'
FAIL: codex_footer_near_miss_async_final_safe_fails_verdict — expected 'VERDICT: NEEDS_REVISION (unrecognized response format — safe-fail)', got 'VERDICT: TIMED_OUT — no response from 'chatgpt-codex-connector[bot]' after 1s (budget 1s) across up to 1 attempt(s)'
PASS: codex_footer_near_miss_async_final_safe_fails_resolves_at_async_final
Tests: 1 passed, 2 failed
```

**Restored, passing run:**

```
PASS: codex_footer_near_miss_async_final_safe_fails_exit_needs_revision
PASS: codex_footer_near_miss_async_final_safe_fails_verdict
PASS: codex_footer_near_miss_async_final_safe_fails_resolves_at_async_final
Tests: 3 passed, 0 failed
```

Scenario that detects the violation: `codex_footer_near_miss_async_final_safe_fails`. (This scenario's mock was already properly isolated — async-final is the terminal site for its branch, so no rescue was possible even before today's fix.)

---

### 4. Decision 6 gate — async-reaction-final, `codex-github-reviewer.sh:2081`

Mutation: same negated-form shape.

```diff
- elif [ "$ASYNC_REACTION_FINAL_BOT_RESPONSE_SOURCE" = "review" ] || ! grep -qi "If Codex has suggestions, it will comment; otherwise it will react with" <<< "$ASYNC_REACTION_FINAL_BOT_RESPONSE_FULL"; then
+ elif ! grep -qi "If Codex has suggestions, it will comment; otherwise it will react with" <<< "$ASYNC_REACTION_FINAL_BOT_RESPONSE_FULL"; then
```

**Baseline (violation absent):**

```
PASS: codex_footer_near_miss_async_reaction_final_safe_fails_exit_needs_revision
PASS: codex_footer_near_miss_async_reaction_final_safe_fails_verdict
PASS: codex_footer_near_miss_async_reaction_final_safe_fails_resolves_at_async_reaction_final
Tests: 3 passed, 0 failed
```

**Failing run (violation present at `codex-github-reviewer.sh:2081`):**

```
FAIL: codex_footer_near_miss_async_reaction_final_safe_fails_exit_needs_revision — expected '1', got '2'
FAIL: codex_footer_near_miss_async_reaction_final_safe_fails_verdict — expected 'VERDICT: NEEDS_REVISION (unrecognized response format — safe-fail)', got 'VERDICT: TIMED_OUT — Codex thumbs-up reaction is not SHA-pinned review evidence (treated as unavailable)'
PASS: codex_footer_near_miss_async_reaction_final_safe_fails_resolves_at_async_reaction_final
Tests: 1 passed, 2 failed
```

**Restored, passing run:**

```
PASS: codex_footer_near_miss_async_reaction_final_safe_fails_exit_needs_revision
PASS: codex_footer_near_miss_async_reaction_final_safe_fails_verdict
PASS: codex_footer_near_miss_async_reaction_final_safe_fails_resolves_at_async_reaction_final
Tests: 3 passed, 0 failed
```

Scenario that detects the violation: `codex_footer_near_miss_async_reaction_final_safe_fails`. (Also already properly isolated — terminal site for the reaction-detected branch.)

---

### 5. Whole-body exact-match classifier (`codex_response_is_approved`) — `codex-github-reviewer.sh:737`

The finding cited `codex-github-reviewer.sh:759` (the `if grep -qE "$template" <<< "$normalized"; then` line inside `codex_response_is_approved`); the template it evaluates lives at line 737. Mutation: remove the trailing `$` anchor from `CODEX_APPROVED_TEMPLATES`' one entry, reopening the "extra trailing content" gap Decision 1 exists to close.

```diff
- ...Try commenting "@codex address that feedback"\. </details>$'
+ ...Try commenting "@codex address that feedback"\. </details>'
```

**Baseline (violation absent):**

```
PASS: codex_e10_trailing_prose_not_approved_exit_needs_revision
PASS: codex_e10_trailing_prose_not_approved_verdict
Tests: 2 passed, 0 failed
```

**Failing run (violation present at `codex-github-reviewer.sh:737`):**

```
FAIL: codex_e10_trailing_prose_not_approved_exit_needs_revision — expected '1', got '0'
FAIL: codex_e10_trailing_prose_not_approved_verdict — expected 'VERDICT: NEEDS_REVISION (unrecognized response format — safe-fail)', got 'VERDICT: APPROVED'
Tests: 0 passed, 2 failed
```

This is a false `APPROVED` — the unsafe direction.

**Restored, passing run:**

```
PASS: codex_e10_trailing_prose_not_approved_exit_needs_revision
PASS: codex_e10_trailing_prose_not_approved_verdict
Tests: 2 passed, 0 failed
```

Scenario that detects the violation: `codex_e10_trailing_prose_not_approved`.

---

### 6. Bounded flavor placeholder — `codex-github-reviewer.sh:737`

Mutation: remove `*` from the placeholder's excluded-character class.

```diff
- ...issues\. [^*`[:cntrl:]]{1,40} \*\*Reviewed commit:\*\*...
+ ...issues\. [^`[:cntrl:]]{1,40} \*\*Reviewed commit:\*\*...
```

**Baseline (violation absent):**

```
PASS: codex_placeholder_asterisk_not_approved_exit_needs_revision
PASS: codex_placeholder_asterisk_not_approved_verdict
Tests: 2 passed, 0 failed
```

**Failing run (violation present at `codex-github-reviewer.sh:737`):**

```
FAIL: codex_placeholder_asterisk_not_approved_exit_needs_revision — expected '1', got '0'
FAIL: codex_placeholder_asterisk_not_approved_verdict — expected 'VERDICT: NEEDS_REVISION (unrecognized response format — safe-fail)', got 'VERDICT: APPROVED'
Tests: 0 passed, 2 failed
```

Also a false `APPROVED`.

**Restored, passing run:**

```
PASS: codex_placeholder_asterisk_not_approved_exit_needs_revision
PASS: codex_placeholder_asterisk_not_approved_verdict
Tests: 2 passed, 0 failed
```

Scenario that detects the violation: `codex_placeholder_asterisk_not_approved`.

---

### Exemption claimed (`REVIEW.md:324`)

The SHA field's `[0-9a-f]{7,40}` bound and the complete footer literal are **not** re-proven here: neither sub-component of the shared template line was modified by this PR's diff (only the flavor-slot substring changed across every commit in this PR) — the SHA bound was already proven by `codex_e5`–`codex_e8` and the footer literal by `codex_e9`–`codex_e11`/`codex_e24` in earlier rounds of this plan. Carrying an already-proven check unmodified is a pure refactor with no behavior change on that surface, per the exemption at `REVIEW.md:324`.

### Full suite after the isolation fix

```
Tests: 722 passed, 0 failed
```

(Unchanged from before — the isolation fix tightens two mocks' conditions without adding or removing any `run_test` assertions.)

See the implementation plan's new **"Planted-violation proof"** section (after "Residual verification strategy") and the smoke-test runbook's new **Step 17** for the indexed, permanent record of this evidence.

File: scripts/development-workflow/codex-github-reviewer.sh [hunk 1]
```diff
@@ -664,21 +623,144 @@ codex_strip_not_only_idiom() {
...
-  local normalized_body
-  normalized_body=$(codex_strip_not_only_idiom "$(codex_strip_quoted_spans "$body")")
-  if grep -qiE "$CODEX_NEGATED_APPROVAL_PATTERN" <<< "$normalized_body"; then
-    return 1
-  fi
-  grep -qiE "$CODEX_APPROVAL_PATTERN" <<< "$normalized_body"
+  local body="$1" normalized template
+  normalized=$(codex_normalize_whitespace "$body")
+  for template in "${CODEX_APPROVED_TEMPLATES[@]}"; do
+    if grep -qE "$template" <<< "$normalized"; then
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply scripts/development-workflow/codex-github-reviewer.sh:759]: Fixed in commit `b4655e6a`: full planted-violation transcripts for all six materially modified checks (the four Decision 6 gate sites, the whole-body classifier's trailing anchor, and the bounded flavor placeholder's excluded-character set) are posted as a PR comment: https://github.com/lhpaul/ai-dev-framework-template/pull/1494#issuecomment-5334816640

Each transcript names the exact file:line, shows the mutation, pastes the actual failing-run output, and pastes the actual restored-passing-run output, naming the scenario that detects the violation.

While producing the transcripts, found and fixed a real pre-existing test-isolation gap: the main-loop and async-arrival near-miss scenarios' mocks returned the same near-miss body on every subsequent poll call, so a broken gate at either site was silently rescued by a downstream site's still-correct gate (visible in the comment's transcript #1, which shows the mutation initially producing no test failure at all before the isolation fix). Both mocks are now call-counted so all four sites are genuinely independent tests.

The evidence is also indexed permanently in the implementation plan's new "Planted-violation proof" section and the smoke-test runbook's new Step 17, not only in this comment.

The SHA field bound and footer literal are explicitly exempted per REVIEW.md:324 (pure refactor, unmodified by this PR's diff), stated rather than silently omitted.

**Comment by lhpaul** [comment]
@codex review (review triggered by workflow runner, commit: b4655e6a2e92)

**Comment by lhpaul** [comment]
## /run-epic PR Disposition

- Scope source: item
- Item: #1491 - codex-github-reviewer.sh: replace open-ended vocabulary enumeration with a conservative verdict classifier
- PR: #1494
- Reviewed head SHA: `b4655e6a`
- Reviewer result: clean
- Blocking findings: 0
- Advisory findings: 0
- Risk: medium
- Merge authority: merge_granted
- Final decision: merge

### Risk Reasons

- workflow script or harness change: scripts/development-workflow/codex-github-reviewer.sh
- workflow guidance change: docs/workflow/development-workflow/protocols/93-automated-reviewer-loop-protocol.md

### Why Safe to Merge

- Scope: Issue #1491 refactor implemented exactly per the merged plan (a0b8f8c7): codex_response_is_approved becomes a whole-body exact-template match; Decision 6 gates the acknowledgement elif on SOURCE != review at all four verdict sites. codex_response_is_blocking is byte-identical to the merge base, verified by function-range diff. Nothing outside the verdict classifier changes behavior.
- Tests: Full harness 722 passed / 0 failed, independently re-run by the orchestrator. Planted-violation proofs recorded for all six modified checks, each citing file:line with both the failing and the restored-passing run. Producing those proofs exposed and fixed two non-diagnostic mocks; all four per-site near-miss scenarios now fail in isolation.
- Reviewer outcome: Step 7a code-reviewer APPROVED after one fix. pr-agent clean; codex-github clean. Post-clean recheck ran with 0 late threads; all review threads resolved.
- CI outcome: Markdown lint, policy, PR-Agent, ShellCheck, workflow test harnesses, CodeRabbit and the reviewer-loop guard all SUCCESS; E2E placeholder SKIPPED. mergeStateStatus CLEAN.
- Rollback or cleanup risk: Single-commit revert restores the prior classifier; no migrations, no persisted state, no external config. Failure direction is safe by design: an unmatched clean response yields NEEDS_REVISION, never a false APPROVED. Live-verified against the real :rocket: and Swish! responses.

### Invocation Policy

- Original command: `/run-item 1491`
- Copy-paste equivalent: `/run-item 1491 --delegate-review --may-merge --may-start-backlog true --max-risk medium --base develop`
- Confirmation: human accepted resolved guardrails policy before mutation (invocation-scoped RUN_ITEM_POLICY_CONFIRMED)

| Field | Recommended | Selected | Effective |
| --- | --- | --- | --- |
| mayStartBacklog |  |  |  |
| delegateReview |  |  |  |
| mayMerge |  |  |  |
| maxRisk |  |  |  |
| base |  |  |  |

### Checkpoint Policy

Not recorded.

### Advisory Decisions

None.

### Security-Sensitive Advisory Findings

None.

### Verification Evidence

- Labels: 
- CI result: unknown
- Reviewer summary present: false
- Unresolved thread count: unknown
- PR merge state: unknown
- Issue state: unknown
- Project status: unknown

### Protocol Deviations

None.
